### PR TITLE
OPERATOR-457 Create equivalent storagecluster object for existing PX

### PR DESCRIPTION
**Pull request by piyush-nimbalkar**
Signed-off-by: Piyush Nimbalkar <pnimbalkar@purestorage.com>

**What this PR does / why we need it**:
- Adds a basic migration workflow. When the operator starts, it starts a goroutine for migration. The migration routine, will first create an equivalent StorageCluster object and wait for the user to approve the migration. 
- The actual migration code is not in the scope of this PR or ticket.
- This covers the migration of just the portworx daemonset. The components are not yet reflected in the StorageCluster object.
- Unit tests and functional tests will be added to this PR soon. 

**Which issue(s) this PR fixes** (optional)
Closes # https://portworx.atlassian.net/browse/OPERATOR-457

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/controller/storagecluster/storagecluster.go
Comment (haibinxie) [inline pkg/controller/storagecluster/storagecluster.go:217]: this should probably be after the validation, so users know if the validation fails or not.

Comment (piyush-nimbalkar) [reply pkg/controller/storagecluster/storagecluster.go:223]: That's a good point. One of the validations that we have marks the storagecluster as failed due to storagecluster being present. I will make changes to it in the subsequent PR where we do actual migration. 
Once that is done, I will move this check after the validation.

File: pkg/controller/storagecluster/storagecluster.go
Comment (pure-jliao) [inline pkg/controller/storagecluster/storagecluster.go:250]: so what happens if the annotation doesn't exist? if understand correctly, during operator install it will construct and create a storage cluster with this annotation set to false?

File: pkg/migration/migration.go
Comment (pure-jliao) [inline pkg/migration/migration.go:161]: feel it's easy to mess up this for loop index, probably could come up with some better ideas to parse them in the future?

File: pkg/migration/migration.go
Comment (pure-jliao) [inline pkg/migration/migration.go:124]: so besides those components, we could construct everything just from portworx container and existing ds object?

**Review by pure-jliao — APPROVED** [review]
Overall looks pretty good to me and assume we'll continue working on this, just have a few questions.

File: pkg/controller/storagecluster/storagecluster.go
Comment (piyush-nimbalkar) [reply pkg/controller/storagecluster/storagecluster.go:250]: Yes, the migration process will create a storagecluster with the annotation set to `false`.

We return `false` only if the annotation is present and set to `false`. That way if this is a migration created cluster, we pause until the user approves it.
If the annotation is absent or has `true` value, we want to go ahead with the normal reconcile loop of creating px pods where absent. I haven't yet figured out how to handle component creation or avoid duplicate components (existing and operator created).

File: pkg/controller/storagecluster/storagecluster.go
Comment (piyush-nimbalkar) [reply pkg/controller/storagecluster/storagecluster.go:251]: In normal scenario, we need to return `false` so it can proceed. 
So when the annotation does not exist, the err will be NotNil. This function will return false and the reconcile will proceed as expected.

File: pkg/migration/migration.go
Comment (piyush-nimbalkar) [reply pkg/migration/migration.go:74]: Yes, but we need to keep checking here too, right? This migration handler will start scaling down the daemonset and then delete it once done. We need to start the scale down/up only when the migration is approved.

File: pkg/migration/migration.go
Comment (piyush-nimbalkar) [reply pkg/migration/migration.go:124]: There are more items missing which are listed on line 448 - https://github.com/libopenstorage/operator/pull/429/files#diff-f343de058945bc383122601555f6ef2a4988c955c6a702167bf6a99c0775fdccR448

File: pkg/migration/migration.go
Comment (piyush-nimbalkar) [reply pkg/migration/migration.go:161]: Yes, I agree. Also, the assumption is that the arguments are always valid. Generally if we pass invalid arguments like arg with no value, non-existent arg, or incorrect value type, oci-mon or portworx will fail to come up. 
This means the existing cluster itself is not up.
But I will see if we can come up with something better and handle more scenarios as we mature this.

Comment (piyush-nimbalkar) [reply pkg/migration/migration.go:161]: Let me know if you have a better idea.

File: pkg/migration/migration.go
Comment (piyush-nimbalkar) [reply pkg/migration/migration.go:71]: In case of error we want to retry the migration as there could be something wrong with k8s api server. The `isMigrationNeeded` returns `true` and empty cluster on error instead of returning the error itself.
Returning the err from `isMigrationNeeded` was fine, but I get the same information from `cluster == nil`. Just wanted to avoid returning an extra argument from that function.

File: pkg/migration/migration.go
Comment (haibinxie) [inline pkg/migration/migration.go:108]: nit: rename startMigration to processMigration

File: pkg/migration/migration.go
Comment (haibinxie) [reply pkg/migration/migration.go:71]: ok literally when I see isMigrationNeeded and true I would think migration is needed, I think return false and and error is easier to understand. Also we dont need to check storagecluster here as it is already created above createStorageClusterIfAbsent right?

File: pkg/migration/migration.go
Comment (haibinxie) [reply pkg/migration/migration.go:74]: If at this point if migration is not approved yet, this will return false and the handler is done. I am not seeing the handler will be called again after it is approved, did I miss anything?

File: pkg/migration/migration.go
Comment (piyush-nimbalkar) [reply pkg/migration/migration.go]: Fixed it by putting this in the retry loop.

File: pkg/migration/migration.go
Comment (piyush-nimbalkar) [reply pkg/migration/migration.go]: Fixed it by putting this in the retry loop.

**Comment by piyush-nimbalkar** [comment]
For some reason, I am not able to reply to some comments. I will reply to them here - 
> ok literally when I see isMigrationNeeded and true I would think migration is needed, I think return false and and error is easier to understand. Also we dont need to check storagecluster here as it is already created above createStorageClusterIfAbsent right?
I removed this and made it simple. If the daemonset is removed only then we stop the migration. If the storagecluster is deleted, we create a new one waiting for approval.

> If at this point if migration is not approved yet, this will return false and the handler is done. I am not seeing the handler will be called again after it is approved, did I miss anything?
When you return false, the outer retry loop continues. It only exits if you return true or an error.

File: pkg/migration/migration.go
Comment (piyush-nimbalkar) [reply pkg/migration/migration.go:66]: I removed this and made it simple. If the daemonset is removed only then we stop the migration. If the storagecluster is deleted, we create a new one waiting for approval.

File: pkg/migration/migration.go
Comment (piyush-nimbalkar) [reply pkg/migration/migration.go:69]: When you return false, the outer retry loop continues. It only exits if you return true or an error.

Comment (haibinxie) [reply pkg/migration/migration.go:69]: oh yes, thought this is out of the poll loop

**Review by haibinxie — APPROVED** [review]

**Pull request merged**
